### ISSUE TITLE
Robust entity handling

### ISSUE THREAD

**Issue by jonathanjfshaw** [jhedstrom/drupalextension#337]
The handling of Drupal entities in Behat Drupal could be consolidated and made more robust.

1) Drupal behat passes complex objects around to represent Drupal entities, back and forth between the driver and extension as well as within each, and these have no predefined structure or interface. As @pfrenssen said let's "transform $entity from a stdClass into a real class".

The entity object we pass around is basically a collection of desired field names and values, and so it gets messy when we went to add non-field values like entity type and these sit amidst the fields with no defined way of distinguishing them, and potentially colliding with them. Ideally we could have some kind of interface that defines properties: entity_type, bundle, id, canonical url, edit url, and fields. So it's $entity->fields that has the collection of desired fields that currently sits at $entity.

2) Originally Drupal behat only supported nodes, terms and users. The new EntityContext (#300) adds support for any other kind of entity. But none of these share any code.  We have 4 separate paths for everything involving creating, cleaning up, and hooking into these entities. Let's create a new entity service that contains generic code for all these things, and a way for specific entity types (both ones we supply like Nodes, Terms and Users as well as arbitrary custom/contrib entities) to extend/replace/override the generic code. This should improve maintainability and extensibility.

**Comment by pfrenssen** [comment]
@jonathanjfshaw are you working on this? I'm willing to give this a go. My idea is to start this from the ground up, in DrupalDriver.

**Comment by jonathanjfshaw** [comment]
I'm not working on this, and to be honest the necessary abstractions are a
bit beyond me. If you created the abstraction framework, the base entity
classes, and the node class, I would be able follow your logic and do terms
and users on the driver.

With some pointers I could probably implement things on the extension side
as well, once you've shown the way driver side.

See how far you get ...

**Comment by ademarco** [comment]
Not sure if it's the same idea you guys had but I was wondering we could kind of "backport" Drupal 8-style entities in the Drupal Driver for Drupal 6 and 7 so that we create a coherent way of dealing with entities across all core versions.

I guess this should be an issue on the DrupalDriver side right? Is there already one?

**Comment by jonathanjfshaw** [comment]
Yes, we should have an issue driver-side, but I don't think we do yet.

And yes, we need objects that live on the driver, but can be addressed from the extension in a version-agnostic way. Which takes us very close to "backporting D8-style entities". 

An important difference though is that even for D8 the objects we need are actually a wrapper around the D8 entity - let's say an "entity handler" object. As I imagine it they would have a $fieldStrings property that held the array of field values that came from the extension, and a $entity property that held the actual Drupal object (entity object in D8, maybe something else more primitive in D6) created based on those fields. Because the contents of the $entity property would vary across core versions, we'd need other properties like $url that were set in version-dependent ways to produce version-independent results; but always mirroring what we'd expect from D8. 

Discovery is the thing that is most complex here. How do developers create additional classes that live outside of the Driver and add support for new entity types, or override the driver's default objects; how to override for just 1 or 2 bundles of an entity type, or just in a specific step or scenario? Particularly how do you do this for a single scenario, using a tag, while using the step definitions from the extension?

My current thinking is to have an "entity manager" service in the driver, instantiated as a property on the rawDrupalContext . This would have a registry of custom entity handlers that you could add classes to in the construct method of your custom context, or in a scenario hook. When you called "$this->entityManager->getNew('node')" it would give you a new node handler object, using your custom handler if you had one currently registered, or the default Driver one if not.

A manual registry like this might seem crude compared to D8's plugin and annotations systems, but it's simple to create and we need something like it to allow for overriding anyway, while on the driver side it allows for discovery of the default handler to happen in exactly the path-based try-and-fallback way it does now for the driver's field handlers.

**Comment by jonathanjfshaw** [comment]
Done right we could also discard all the complex code for hooking into @beforeUserCreate.
If people want to do things like that, they'd just create a custom User handler extending the provided one and add their own logic in it's create() method.

**Comment by AlexSkrypnyk** [comment]
## Narrowing scope to a minimal `EntityStub`

After re-reading this thread against the current state of the code (driver
`v3` capability interfaces, `D10`/`D11`-only support, attribute-based hooks,
sub-contexts), most of the original 2017 proposal is no longer the right
shape for the codebase. We need to cut down significantly on the proposed
functionality and land only the small piece that still has real value.

### What's no longer needed

| 2017 proposal                                       | Why it's out |
|-----------------------------------------------------|--------------|
| Entity-handler registry on the driver               | Capability interfaces already separate driver-level concerns |
| Per-bundle / per-scenario handler subclasses        | Behat hooks + sub-contexts already provide this extension point |
| Per-version handler subclasses (`D6`/`D7`/`D8`)     | Driver `v3` is `D10`/`D11`-only, no cross-version wrapping needed |
| Backporting `D8` entities to `D6`/`D7`              | `D7` is EOL since Jan 2025 |
| Typed subclasses (`NodeStub`, `TermStub`, ...)      | Re-introduces the "parallel Drupal entity system" smell - one class mirrors Drupal Core's `Entity::create($type, $values)` shape |
| `nakedEntityCreate` style helpers per entity type   | Generic `entityCreate(string $type, ...)` already exists in `ContentCapabilityInterface` |

### What's left, and what's actually worth doing

A single typed `EntityStub` class plus consolidating the four
`nodeCreate`/`termCreate`/`userCreate`/`languageCreate` paths in
`RawDrupalContext` into one `entityCreate()`. Concretely:

- One `final` class, type and bundle as `readonly`, values as a mutable bag,
  plus a slot for the saved Drupal entity object after the driver returns.
- Capability interfaces typed against `EntityStub` instead of `\stdClass`.
- `RawDrupalContext::entityCreate(EntityStub)` replaces the four per-type
  methods. The four public methods stay as thin BC wrappers, deprecated for
  removal in a future major.
- Single `$createdStubs` array replaces the four `$nodes` / `$terms` /
  `$languages` cleanup arrays. `cleanEntities()` walks them in reverse so
  dependent entities come down before their dependencies.
- The two `instanceof \stdClass` runtime guards in `RawDrupalContext` get
  deleted - the type system enforces the contract now.

### What this gives us

The DX win is bounded but real: IDE autocomplete on the stub, `PHPStan` can
catch typos against declared properties, refactor-rename works,
field-vs-base-property collisions go away by structure (metadata on the
class, values in `$stub->getValues()`), and the runtime null-checks
masquerading as type guards are replaced by actual types.

What this does **not** give: type safety on the values bag (it's still
`mixed`), validation, or encapsulation of save logic. None of those are
needed - the stub is a data envelope, nothing more.

### Driver-side coordination

The driver is the lower layer and a dependency of the extension. The
`EntityStub` class lives in `Drupal\Driver\EntityStub` (in
`drupal/drupal-driver`) so the capability interfaces can type against it
without inverting the dependency direction.

Driver-side issue:
[`jhedstrom/DrupalDriver#362`](https://github.com/jhedstrom/DrupalDriver/issues/362).
That lands first in the in-progress `drupal/drupal-driver` `3.x` line, then
this extension repo bumps the constraint and does the consumer-side
refactor.

### Sequencing

1. `drupal/drupal-driver` `3.x`: introduce `EntityStub`, retype capability
   interfaces, ship a transition release that accepts both `EntityStub` and
   `\stdClass` with the latter deprecated.
2. This repo (`v6.0.0`): bump the driver constraint to the release that
   lands `EntityStub`, replace `\stdClass` flows with `EntityStub`,
   consolidate the four `*Create()` paths, regenerate `STEPS.md`.

### Suggested action on this issue

Keep this issue open as the umbrella tracking the extension-side work, but
update the title and description to reflect the narrowed scope. The
driver-side work is now tracked in
[`jhedstrom/DrupalDriver#362`](https://github.com/jhedstrom/DrupalDriver/issues/362).
The original 2017 proposal of an entity-handler registry, per-version
subclasses, and per-bundle handler discovery is closed-out as superseded
by capability interfaces and Behat's existing extension points.

**Comment by AlexSkrypnyk** [comment]
## Resolved

Landed with the narrowed scope agreed on the
[earlier comment](https://github.com/jhedstrom/drupalextension/issues/337#issuecomment-4343081519):
a single typed `EntityStub` replaces `\stdClass`, the four
`nodeCreate` / `termCreate` / `userCreate` / `languageCreate` paths in
`RawDrupalContext` are consolidated into one `entityCreate()` and one
`cleanEntities()` cleanup hook, and the hook scopes plus the
user / authentication managers are retyped to take
`EntityStubInterface`. The 2017 ideas around an entity-handler registry,
per-bundle handler subclasses and per-version wrapping were dropped as
superseded by capability interfaces and the existing Behat hook +
sub-context mechanism.

### Driver side

`drupal/drupal-driver`:

- Issue: [`jhedstrom/DrupalDriver#362`](https://github.com/jhedstrom/DrupalDriver/issues/362)
- PR: [`jhedstrom/DrupalDriver#363`](https://github.com/jhedstrom/DrupalDriver/pull/363)

Adds `Drupal\Driver\Entity\EntityStubInterface` and the final
`EntityStub` implementation, retypes every capability interface
(`Content`, `User`, `Authentication`, `Language`, `Block`) and the
concrete drivers to flow `EntityStubInterface` instead of `\stdClass`.

### Extension side

`drupal/drupal-extension`:

- PR: [#814](https://github.com/jhedstrom/drupalextension/pull/814)

Bumps the driver constraint, replaces every `\stdClass` flow with
`EntityStubInterface`, consolidates the four `*Create()` paths and the
four cleanup hooks, retypes `BaseEntityScope` (`getEntity()` →
`getStub()`) and the user / authentication managers, and drops the
runtime `instanceof \stdClass` guards.

Closing this issue.
